ouroborus/lib/ouroborus/container.rb

```diff
@@ -0,0 +1,171 @@
+      return normalize_arg arg.to_s unless arg.is_a? String
+      return "''" if arg.empty?
+      arg.chars.reduce '' do | acc, curr |
+        next "#{acc}\\#{curr}" if backslasheable? curr
+        "#{acc}#{curr}"
+      end
+    end
+
+    def backslasheable? c
+      ' \\$#\'\"'.include? c
+    end
+  end
+  class RestartCondition
+    class << self
+      STATUSES = {
+        :ALWAYS => "always",
+        :NEVER => "no",
+        :UNLESS_STOPPED => "unless-stopped",
+        :ON_FAILURE => "on-failure"
+      }
+      def valid?(condition)
+        STATUSES.has_key? condition
+      end
+      def as_string(condition)
+        STATUSES[condition]
+      end
+    end
+  end
+  class Container
+    attr_reader :args
+    def initialize(name:, image:, tag: 'latest')
+      @name = name
+      @image = image
+      @tag = tag
+      @dockerArgs = Args.new
+      @args = Args.new
+      @daemon = true
+    end
+
+    def daemon?
+      @daemon
+    end
+
+    def daemon=(value)
+      if value
+        @daemon = true
+      else
+        @daemon = false
+      end
+    end
+
+    def imgTag
+      "#{@image}:#{@tag}"
+    end
+
+    def to_s
+      replaceable = Args.new
+      replaceable << "docker" << as_args
+      "#{replaceable}"
+    end
+
+    def as_args
+      replaceable = Args.new
+      replaceable << "run"
+      if daemon?
+        replaceable << "-d"
+      else
+        replaceable << "-i"
+      end
+      replaceable << @dockerArgs << "--name" << @name << imgTag << @args
+    end
+
+    def port(incoming, interface = nil)
+      p = "#{incoming}"
+      p += ":#{interface}" unless interface.nil?
+      @dockerArgs << '-p' << p
+    end
+
+    def volume(their, ours = nil)
+      v = "#{their}"
+      v += ":#{ours}" unless ours.nil?
+      @dockerArgs << '-v' << v
+    end
+
+    def restart(condition, maxRetries = nil)
+      raise "restart condition must met RestartCondition.valid?" unless RestartCondition.valid? condition
+      condStr = if !maxRetries.nil?
+        raise "restart maxRetries should only be set when condition is :ON_FAILURE" unless condition == :ON_FAILURE
+        "#{RestartCondition.as_string condition}:#{maxRetries}"
+      else
+        "#{RestartCondition.as_string condition}"
+      end
+      @dockerArgs << '--restart' << condStr
+    end
+
+    def env(name, value = nil)
+      @dockerArgs << '--env' << if name.is_a? Hash
+        raise "When first arg is a hash, env second arg should be nil" unless value.nil?
+        name.map do |kv|
+          normalizeEnvValue(*kv)
+        end.reduce("") do |acc, envvalue|
+          if !acc.empty?
+            acc + "," + envvalue
+          else
+            envvalue
+          end
+        end
+      elsif name.is_a? Array
+        usedValue = if value.nil? then
+          []
+        elsif value.is_a? Array then
+          usedValue = value
+        else
+          raise "When first arg is an Array, env second arg must be another array or nil"
```

Comment:
Just like above: raise "When first arg is an Array, env second arg must be another array or __should not be passed__"

Reply:
```suggestion
          raise "When first arg is an Array, env second arg must be another array or should not be passed"
```